### ISSUE TITLE
level=error msg="Handler for GET /v1.17/containers/4921160fdc99796b2a1cd133b23bff6a9e77bfd7971f537a2cdfd7b9fbe35e6b/logs returned error: Failed to get logging factory: logger: no log driver named 'none' is registered"

### ISSUE THREAD

**Issue by gswallow** [docker/docker#29915]
Our log collection container uses the docker logs API through /var/run/docker.sock to collect container logs.  Apparently, API v1.17 (in docker 1.11.2) can't handle containers that get run with the --log-driver=none flag.

/etc/systemd/system/ecs-agent.service:

```
root@ip-172-20-176-161:~# cat /etc/systemd/system/ecs-agent.service 
[Unit]
Description=AWS ECS Agent
Documentation=https://docs.aws.amazon.com/AmazonECS/latest/developerguide/
Requires=docker.socket
After=docker.service

[Service]
Environment=ECS_VERSION=v1.13.1
Restart=on-failure
RestartSec=30
RestartPreventExitStatus=5
SyslogIdentifier=ecs-agent
ExecStartPre=-/usr/bin/docker kill ecs-agent
ExecStartPre=-/usr/bin/docker rm ecs-agent
ExecStartPre=/usr/bin/docker pull amazon/amazon-ecs-agent:${ECS_VERSION}
ExecStart=/usr/bin/docker run --name ecs-agent \
 --env-file=/etc/ecs/ecs.config \
 --volume=/var/run/docker.sock:/var/run/docker.sock \
 --volume=/var/log/ecs:/log \
 --volume=/var/lib/ecs/data:/data \
 --volume=/sys/fs/cgroup:/sys/fs/cgroup:ro \
 --volume=/run/containerd:/run/containerd:ro \
 --log-driver=journald \
 --net=host \
 --publish=127.0.0.1:51678:51678 \
 --env=ECS_LOGFILE=/log/ecs-agent.log \
 --env=ECS_DATADIR=/data \
 --env=ECS_APPARMOR_CAPABLE=true \
 amazon/amazon-ecs-agent:${ECS_VERSION}

ExecStopPost=/usr/bin/docker stop -t 2 ecs-agent
ExecStopPost=/usr/bin/docker rm ecs-agent

[Install]
WantedBy = multi-user.target
```

/etc/systemd/system/sumologic-collector.service:

```
[Unit]
Description=SumoLogic Collector
Documentation=http://help.sumologic.com/
Requires=syslog.socket
After=network.target

[Service]
Environment=SUMOLOGIC_VERSION=latest-file
Restart=on-failure
RestartSec=30
RestartPreventExitStatus=5
SyslogIdentifier=sumologic-collector
ExecStartPre=-/usr/bin/docker kill sumologic-collector
ExecStartPre=-/usr/bin/docker rm sumologic-collector
ExecStartPre=/usr/bin/docker pull sumologic/collector:latest-file
ExecStart=/usr/bin/docker run --name sumologic-collector \
 --env-file=/etc/sumo/sumologic.conf \
 --volume=/var/lib/docker:/logs:ro \
 --volume=/var/log:/hostlogs:ro \
 --volume=/var/run/docker.sock:/var/run/docker.sock \
 --volume=/etc/sumo/sumo-sources.json:/etc/sumo-sources.json:ro \
 --restart=always \
 --cap-add=SYS_PTRACE \
 --security-opt=apparmor:unconfined \
 sumologic/collector:${SUMOLOGIC_VERSION}

ExecStopPost=/usr/bin/docker stop -t 2 sumologic-collector
ExecStopPost=/usr/bin/docker rm sumologic-collector

[Install]
WantedBy = multi-user.target
```

/etc/sumo/sumo-sources.json:

```
{
  "api.version": "v1",
  "sources": [
    {
      "sourceType": "LocalFile",
      "name": "localfile-var-log-syslog",
      "pathExpression": "/hostlogs/syslog*",
      "multilineProcessingEnabled": false,
      "automaticDateParsing": true,
      "forceTimeZone": false,
      "category": "qa1/syslog"
    },
    {
      "sourceType": "LocalFile",
      "name": "localfile-var-log-authlog",
      "pathExpression": "/hostlogs/auth.log*",
      "multilineProcessingEnabled": false,
      "automaticDateParsing": true,
      "forceTimeZone": false,
      "category": "qa1/authlog"
    },
    {
      "sourceType": "LocalFile",
      "name": "localfile-var-log-ecs-ecs-agentlog",
      "pathExpression": "/hostlogs/ecs/ecs-agent.log*",
      "multilineProcessingEnabled": false,
      "automaticDateParsing": true,
      "forceTimeZone": false,
      "category": "qa1/ecs-agent-logs"
    },
    {
      "name": "Docker-logs",
      "category": "qa1/docker-logs",
      "allContainers": true,
      "collectEvents": true,
      "uri": "unix:///var/run/docker.sock",
      "specifiedContainers": [],
      "multilineProcessingEnabled": false,
      "sourceType": "DockerLog"
    }
  ]
}
```

Constant error logs (like, 9gb in less than one day):

```
Jan  5 18:13:55 ip-172-20-176-161 docker[12315]: time="2017-01-05T18:13:55.284195400Z" level=error msg="Handler for GET /v1.17/containers/4921160fdc99796b2a1cd133b23bff6a9e77bfd7971f537a2cdfd7b9fbe35e6b/logs returned error: Failed to get logging factory: logger: no log driver named 'none' is registered"
Jan  5 18:13:55 ip-172-20-176-161 docker[12315]: time="2017-01-05T18:13:55.285742252Z" level=error msg="Handler for GET /v1.17/containers/4921160fdc99796b2a1cd133b23bff6a9e77bfd7971f537a2cdfd7b9fbe35e6b/logs returned error: Failed to get logging factory: logger: no log driver named 'none' is registered"
Jan  5 18:13:55 ip-172-20-176-161 docker[12315]: time="2017-01-05T18:13:55.287292358Z" level=error msg="Handler for GET /v1.17/containers/4921160fdc99796b2a1cd133b23bff6a9e77bfd7971f537a2cdfd7b9fbe35e6b/logs returned error: Failed to get logging factory: logger: no log driver named 'none' is registered"
Jan  5 18:13:55 ip-172-20-176-161 docker[12315]: time="2017-01-05T18:13:55.288698137Z" level=error msg="Handler for GET /v1.17/containers/4921160fdc99796b2a1cd133b23bff6a9e77bfd7971f537a2cdfd7b9fbe35e6b/logs returned error: Failed to get logging factory: logger: no log driver named 'none' is registered"
```

It would be nice to see a backported fix for this, since I can't upgrade to v1.12.x due to other API bugs/backward incompatibility/whatever that pretty much break AWS ECS.

**Comment by gswallow** [comment]
Oh, hell.  So I configured the container I don't want logs from to use the syslog driver instead, and now I get this:

```
Jan  5 19:22:55 ip-172-20-176-161 docker[1117]: time="2017-01-05T19:22:55.894198023Z" level=error msg="Handler for GET /v1.17/containers/a8ba3caf03a2d033a78b53aa32d5de2adc6461208496c422ff76bf2b0e9aad51/logs returned error: configured logging reader does not support reading"
```

**Comment by cpuguy83** [comment]
@gswallow Not sure what you are expecting to happen here.
We do catch the "none" error earlier now, but you'll still get an error when you tried to read logs on none.
Likewise with log drivers that don't support reading.

Currently the only drivers that do support reads are jsonfile and journald.
This is generally because other protocols, e.g. syslog, do not have a standard way to retrieve logs.

**Comment by gswallow** [comment]
My intent here was to separate the ECS agent container's logs from everything else.  The ECS agent container logs to stdout, which gets collected by Docker in /var/lib/container/X/log, and it logs to a file mapped as a volume: /var/log/ecs/ecs-agent.log*.  I added --log-driver=none for that container in its systemd file.

Docker shouldn't log errors to journald when it finds a container that has been explicitly told not to log?

**Comment by cpuguy83** [comment]
The error in this case is the user requested something that is impossible -- e.g. To consume logs on a container that doesn't support this.

I agree we probably shouldn't log errors which are user errors. Today we log every non-successful request as an error, though.

**Comment by thaJeztah** [comment]
Looks like your settings attempt to get logs for all containers (`allContainers: true`), probably need to narrow that down to only attempt getting logs from containers that have logging enabled

**Comment by gswallow** [comment]
If I understand the docker remote API docs correctly, one (in this case, Sumologic) would need to get a list of the containers first by GETting /containers, and then for each container ID or name in the list, GET /containers/<id>/logs, optimally with a timestamp.

With docker remote API v1.19 or earlier, the only way to filter the first call (get all containers) is by exit code, status, or label.  I suppose Sumologic could get the list of containers and then filter out containers whose tags include "ignore-my-logs," or something.  But the filtering logic would need to be written by the Sumologic guys.

**Comment by thaJeztah** [comment]
Yes, I think something like that would be needed. I'm not familiar with Sumologic, so don't know what options they offer.

Another approach is to not "poll" the docker remote API to get those logs; if you want to centralise logging, use a logging driver that sends all logs to a central logging collector, https://docs.docker.com/engine/admin/logging/overview/

> I agree we probably shouldn't log errors which are user errors. Today we log every non-successful request as an error, though.

I'm not sure what's best here; is "silently" ignoring the problem the best solution?

**Comment by gswallow** [comment]
I would agree about what's best.  It should be obvious to the docker engine that I configured a container with --log-driver=none.  

What is your policy on backporting changes to older docker releases?  We're sort of stuck on version 1.11.x because AWS ECS and Empire recommend/require it, but the fix, mentioned by @cpuguy83, above, applies to Docker 1.12.4 onwards.  My solution now is just to *discard* AWS ECS logs lower than "crit" priority.  The ECS agent is very chatty and its logs are less than helpful, but telling the ECS agent to only log critical messages may bite me in the end.

**Comment by thaJeztah** [comment]
> It should be obvious to the docker engine that I configured a container with --log-driver=none.

Well, it is, so it outputs an error if someone still tries to obtain logs through the API

> What is your policy on backporting changes to older docker releases

We maintain only the "current" release; no backporting.

**Comment by gswallow** [comment]
That's unfortunate.  Unrelated changes in the API mean I can't upgrade.
Hopefully you'll take that step towards product maturiity soon.

**Comment by thaJeztah** [comment]
> Unrelated changes in the API mean I can't upgrade.

Could you add more information about this? The API is versioned, so docker 1.12 or 1.13 still supports the same API version as 1.11 uses (API version 1.23); https://docs.docker.com/engine/reference/api/docker_remote_api/

**Comment by gswallow** [comment]
That's good to know.  I know that I'm being affected by this compatibility issue: https://github.com/remind101/empire/pull/915, which was fixed after the release of Empire 0.10.1.  Because the Remind team switched their default backend scheduler, I'll need to test the migration to Empire 0.11.0 before we can adopt it in production.

I think I'll just wait until both Empire and the ECS agent recommend a newer version of Docker.

**Comment by thaJeztah** [comment]
Ah, yes, the `/containers/<id>/copy` endpoint was deprecated in docker 1.8, and does not work with API versions 1.24 and up; [docs/deprecated.md#containersid-or-namecopy-endpoint](https://github.com/docker/docker/blob/v1.12.0/docs/deprecated.md#containersid-or-namecopy-endpoint) (https://github.com/docker/docker/pull/22149). It should still work on API 1.23 and under though. If that's not the case, please open a bug report
